// backend/dist/bd.js
import mysql from 'mysql2/promise';
import dotenv from 'dotenv';
// Carrega as variáveis de ambiente do arquivo .env
// Isso garante que as credenciais não fiquem hardcoded no código
dotenv.config();
/**
 * Cria e retorna a configuração do banco de dados a partir de variáveis de ambiente
 * Se alguma variável não estiver definida, usa valores padrão seguros
 */
function getDatabaseConfig() {
    return {
        host: process.env.DB_HOST || 'localhost',
        port: parseInt(process.env.DB_PORT || '3306', 10),
        user: process.env.DB_USER || 'root',
        password: process.env.DB_PASSWORD || '',
        database: process.env.DB_NAME || 'dados',
        waitForConnections: true,
        connectionLimit: parseInt(process.env.DB_POOL_MAX || '10', 10),
        queueLimit: 0, // Sem limite de fila
        idleTimeout: parseInt(process.env.DB_POOL_IDLE_TIMEOUT || '30000', 10),
        acquireTimeout: parseInt(process.env.DB_POOL_ACQUIRE_TIMEOUT || '60000', 10),
    };
}
/**
 * Pool de conexões MySQL
 *
 * POR QUE USAR POOL?
 * - Evita criar/fechar conexões a cada operação (mais eficiente)
 * - Reutiliza conexões existentes (melhor performance)
 * - Gerencia automaticamente o número de conexões simultâneas
 * - Reduz overhead de estabelecer novas conexões
 *
 * O pool é criado uma única vez e reutilizado em todas as operações
 */
let pool = null;
/**
 * Obtém ou cria o pool de conexões do banco de dados
 *
 * SINGLETON PATTERN: Garante que apenas um pool seja criado
 * Isso é importante porque criar múltiplos pools pode esgotar recursos
 *
 * @returns Pool de conexões MySQL configurado
 * @throws Error se não conseguir criar o pool
 */
export function getPool() {
    if (!pool) {
        try {
            const config = getDatabaseConfig();
            // Validação básica: verifica se o nome do banco foi informado
            if (!config.database) {
                throw new Error('Nome do banco de dados (DB_NAME) não foi configurado no arquivo .env');
            }
            pool = mysql.createPool(config);
            // Event listeners para monitoramento e debug
            pool.on('connection', (connection) => {
                console.log(`[DB] Nova conexão estabelecida (ID: ${connection.threadId})`);
            });
            pool.on('error', (err) => {
                console.error('[DB] Erro no pool de conexões:', err);
                // Se o erro for de conexão perdida, tenta reconectar
                if (err.code === 'PROTOCOL_CONNECTION_LOST') {
                    console.warn('[DB] Tentando reconectar ao banco de dados...');
                    pool = null; // Força recriação do pool na próxima chamada
                }
            });
            console.log('[DB] Pool de conexões criado com sucesso');
        }
        catch (error) {
            console.error('[DB] Erro ao criar pool de conexões:', error);
            throw new Error(`Falha ao conectar ao banco de dados: ${error instanceof Error ? error.message : 'Erro desconhecido'}`);
        }
    }
    return pool;
}
/**
 * Obtém uma conexão do pool para executar queries
 *
 * IMPORTANTE: Sempre libere a conexão após o uso chamando connection.release()
 * ou use try/finally para garantir que a conexão seja liberada mesmo em caso de erro
 *
 * @returns Promise com uma conexão do pool
 */
export async function getConnection() {
    const poolInstance = getPool();
    try {
        const connection = await poolInstance.getConnection();
        return connection;
    }
    catch (error) {
        console.error('[DB] Erro ao obter conexão do pool:', error);
        throw new Error(`Não foi possível obter conexão do banco de dados: ${error instanceof Error ? error.message : 'Erro desconhecido'}`);
    }
}
/**
 * Fecha o pool de conexões
 *
 * Use esta função apenas quando for encerrar a aplicação
 * (ex: em um handler de shutdown graceful)
 */
export async function closePool() {
    if (pool) {
        await pool.end();
        pool = null;
        console.log('[DB] Pool de conexões fechado');
    }
}
/**
 * Função de compatibilidade com código antigo
 * @deprecated Use getPool() ou getConnection() ao invés desta função
 * Esta função ainda funciona, mas não é recomendada para novos códigos
 */
export async function connection() {
    return getConnection();
}
